### ISSUE TITLE
Cluster size for moco v2 + CLD seems to be too big?

### ISSUE THREAD

**Issue by chengxuz** [frank-xwang/CLD-UnsupervisedLearning#11]
Hi,

Thanks for this interesting work. When I was running the training script for moco v2 + CLD, I observed that the effective batch size (64=512/8) is actually smaller than the cluster size you assigned (120). This makes the clustering trivial, e.g. every example is one cluster. Can I confirm that this is not a bug? Or did I run it in the wrong way?

**Comment by frank-xwang** [comment]
Hi Chengxu, thanks for your interest in our work! For the selection of cluster size, we have provided some discussions in the paper. The answer is based on our discussion in the paper and some new information, hope it can answer your question: The ideal number of groups depends on the level of instance correlation, the number of classes, and the mini-batch size. Figure 3 in the paper shows that for high-correlation dataset Kitchen-HC, CLD is best when the number of groups is close to the number of classes, i.e., 10, although CLD still outperforms NPID/MoCo at 80 groups. For ImageNet, the instance correlation is low. Also, the number of classes of 1,000 is much larger than the batch size our 8 GPUs can afford. Therefore, we just choose the largest number of groups possible. We expect continuous gain with more groups and larger batches afforded by more GPUs. We also tried some local experiments with a much larger batch size on a very small network, we observed consistent performance gains. Recently, I also saw some papers that tried to avoid this problem by applying clustering or nearest neighboring on the memory-bank/queue, which has much more prototypes than the number of classes and has a high chance to contain high correlated prototypes.

To answer your question in full, CLD gains on four aspects: 1) Integrate between-instance similarity into contrastive learning via cross-level discrimination;2) Impose invariant mapping thru cross-view comparisons;3) Improve the positive/negative sample ratio of contrastive learning and hence invariant mapping with batch-wise and cross-view comparisons; 4) Reduce feature variance from data  augmentation with proposed normalized projection head. Aspects 2, 3, and 4 remain valid even as the largest number of groups possible is chosen. We expect an even larger gain with moregroups and larger batches afforded by more GPUs. Nevertheless, CLD wins by merits instead of a large compute.
